src/main/java/graphql/schema/transform/FieldVisibilitySchemaTransformation.java

```diff
@@ -0,0 +1,206 @@
+                    element, context.getParentNode());
+            if (!visibilityPredicate.isVisible(environment)) {
+                deleteNode(context);
+
+                if (element instanceof GraphQLFieldDefinition) {
+                    removedTypes.add(((GraphQLFieldDefinition) element).getType());
+                } else if (element instanceof GraphQLInputObjectField) {
+                    removedTypes.add(((GraphQLInputObjectField) element).getType());
+                }
+            }
+
+            return TraversalControl.CONTINUE;
+        }
+    }
+
+    private static class TypeVisibilityVisitor extends GraphQLTypeVisitorStub {
+
+        private final Set<String> protectedTypeNames;
+        private final Set<GraphQLType> observedBeforeTransform;
+        private final Set<GraphQLType> observedAfterTransform;
+        private final Set<GraphQLType> removedTypes;
+
+        private TypeVisibilityVisitor(Set<String> protectedTypeNames,
+                                      Set<GraphQLType> observedTypes,
+                                      Set<GraphQLType> observedAfterTransform,
+                                      Set<GraphQLType> removedTypes) {
+            this.protectedTypeNames = protectedTypeNames;
+            this.observedBeforeTransform = observedTypes;
+            this.observedAfterTransform = observedAfterTransform;
+            this.removedTypes = removedTypes;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLInterfaceType(GraphQLInterfaceType node,
+                                                          TraverserContext<GraphQLSchemaElement> context) {
+            return super.visitGraphQLInterfaceType(node, context);
+        }
+
+        @Override
+        public TraversalControl visitGraphQLType(GraphQLSchemaElement node,
+                                                 TraverserContext<GraphQLSchemaElement> context) {
+            if (observedBeforeTransform.contains(node) &&
+                    !observedAfterTransform.contains(node) &&
+                    (node instanceof GraphQLObjectType ||
+                            node instanceof GraphQLEnumType ||
+                            node instanceof GraphQLInterfaceType ||
+                            node instanceof GraphQLUnionType)) {
+
+                return deleteNode(context);
+            }
+
+            return TraversalControl.CONTINUE;
+
+//            if (!observedBeforeTransform.contains(node) &&
+//                    node.getInterfaces().stream().noneMatch(observedBeforeTransform::contains) &&
+//                    node.getInterfaces().stream().anyMatch(removedTypes::contains) &&
+//                    !ScalarInfo.isStandardScalar(node.getName()) &&
+//                    !protectedTypeNames.contains(node.getName())) {
+//                return deleteNode(context);
+//            }
+//
+//            return TraversalControl.CONTINUE;
+        }
```

Comment:
Can this commented out code be removed ??

Reply:
Removed.